### DISCUSSION TITLE
Use IntoFuture for Mutex.lock

### DISCUSSION THREAD

**Discussion by LasseRosenow** [tokio-rs/tokio#5047] [Ideas] [answered]
Rust 1.64 introduced the IntoFuture Trait. with this we could write:

```rust
let mutex = Mutex::new(0);
mutex.await;
```

instead of 
```rust
let mutex = Mutex::new(0);
mutex.lock().await;
```

**Comment by Darksonn** [comment, marked as answer]
Ignoring the fact that this is incompatible with our minimum-supported-rust-version, that doesn't seem like a good idea.  It seems less confusing to have an explicit `lock` method.

Also, I believe the `.await` of `IntoFuture` consumes the argument, which is not appropriate for locking a mutex.

**Reply by hawkw** [reply]
Additionally, this would be an API change that's somewhat inconsistent with `std::sync::Mutex` --- you don't lock a mutex by "calling" the mutex as though it were a function, you call its `lock` method. The use of `IntoFuture` here would be worth considering in a hypothetical world where `std::sync::Mutex`'s API looked like this:

```rust
// if you could write this...
let mutex = std::sync::Mutex::new(0);
let _guard = mutex();

// ...then it might make more sense to write this
let mutex = tokio::sync::Mutex::new(0);
let _guard = mutex.await;
```
but since that's not the case, it's probably better to have a `lock()` method for consistency.
